### ISSUE TITLE
Odroid N2 and FFmpeg Hardware Acceleration

### ISSUE THREAD

**Issue by dziobson** [blakeblackshear/frigate#901]
Hi
Whats the best ffmpeg configuration for that kind of board ?

**Comment by spattinson** [comment]
I am looking to do this too. I have got as far as confirming that the version of ffmpeg bundled in frigate only supports h264_v4l2m2m  acceleration for pi4 and that doesn't work on my rk3399 based board , and possibly doesn't work for odroid n2 either
```root@ccab4aaf-frigate:~# ffmpeg -hwaccels
ffmpeg version 4.3.1 Copyright (c) 2000-2020 the FFmpeg developers
  built with gcc 9 (Ubuntu 9.3.0-17ubuntu1~20.04)
  configuration: --disable-debug --disable-doc --disable-ffplay --enable-shared --enable-avresample --enable-libopencore-amrnb --enable-libopencore-amrwb --enable-gpl --enable-libfreetype --enable-libvidstab --enable-libmp3lame --enable-libopus --enable-libtheora --enable-libvorbis --enable-libvpx --enable-libwebp --enable-libxcb --enable-libx265 --enable-libxvid --enable-libx264 --enable-nonfree --enable-openssl --enable-libfdk_aac --enable-postproc --enable-small --enable-version3 --enable-libzmq --extra-libs=-ldl --prefix=/opt/ffmpeg --enable-libopenjpeg --enable-libkvazaar --enable-libaom --extra-libs=-lpthread --enable-v4l2_m2m --enable-neon --extra-cflags=-I/opt/ffmpeg/include --extra-ldflags=-L/opt/ffmpeg/lib
  libavutil      56. 51.100 / 56. 51.100
  libavcodec     58. 91.100 / 58. 91.100
  libavformat    58. 45.100 / 58. 45.100
  libavdevice    58. 10.100 / 58. 10.100
  libavfilter     7. 85.100 /  7. 85.100
  libavresample   4.  0.  0 /  4.  0.  0
  libswscale      5.  7.100 /  5.  7.100
  libswresample   3.  7.100 /  3.  7.100
  libpostproc    55.  7.100 / 55.  7.100
Hardware acceleration methods:

root@ccab4aaf-frigate:~#
```
I am not well versed with docker and was wondering how to rebuild with a ffmpeg that supports more hwaccel options. I have a rockchip rk3399 based board which has a gpu that may be able to do some hw acceleration, I have compiled ffmpeg with support for vdpau, vaapi, drm, and support for the rkmpp hardware decoder.
These are the extra build options:
```
--enable-vdpau \
 --enable-vaapi \
 --enable-libdrm \
 --enable-rkmpp \
```

**Comment by spattinson** [comment]
30% of one core or total? I think what happens is if you get hw decode working it may be that there is not zero copy so there us still a lot of data being copied around. I got 25% of one core on nanopc t4 rk3399 on mainline and briefly 5% on legacy 4.18x bsp kernel. Nether were reliable, crap out after a short time. Most effort seems to be to decoding and rendering for the likes of kodi. 

Some significant work went in to the latest 5.13 kernel for hantro and rkvdec decoders in rk3399. I feel that ffmpeg needs to sync up with that. I have used my sons old pc in mean time with an old nvidia card nvdec. Its rock solid in 8yr old intel and uses 5% of one core per cam. Theres a lot of hate in linux community for nvidia for not open sourcing, but there drivers just work without hassle in most cases

**Comment by baldisos** [comment]
Thats 30% across all cores in total. I'm using the HA Addon. When looking at the Addon-Status it says the Addon itself takes up about 21.5% CPU load.

**Comment by hkrob** [comment]
> Are there any tips on HW-Acceleratoion for the N2+? Although i could bring the CPU usage down from 60% to 30% by using ffmpeg args, i'm still not satisfied with that.

Mind sharing the args you are using? Or have you found something better for hwaccel on the N2?

**Comment by baldisos** [comment]
> 
> 
> > Are there any tips on HW-Acceleratoion for the N2+? Although i could bring the CPU usage down from 60% to 30% by using ffmpeg args, i'm still not satisfied with that.
> 
> Mind sharing the args you are using? Or have you found something better for hwaccel on the N2?

Do you mean the input args used?

```
      input_args:
        - -avoid_negative_ts
        - make_zero
        - -fflags
        - nobuffer
        - -flags
        - low_delay
        - -strict
        - experimental
        - -fflags
        - +genpts+discardcorrupt
        - -rw_timeout
        - "5000000"
        - -use_wallclock_as_timestamps
        - "1"
```

**Comment by hkrob** [comment]
Ah ok, thanks.. I'm using about the same and still seeing quite high CPU.

I guess I need to switch to an intel quick sync box

Thanks for sharing

**Comment by bdherouville** [comment]
Hi guys !

I modified and tested the frigate ffmpeg docker container. CPU is at 2% for full HD video.

Can you test and share your thoughts ?

Regards

https://github.com/blakeblackshear/frigate/pull/1814

**Comment by hkrob** [comment]
> Hi guys !
> 
> I modified and tested the frigate ffmpeg docker container. CPU is at 2% for full HD video.
> I have a RockPro64 4Gb.
> 
> Can you test and share your thoughts ?
> 
> Regards
> 
> https://github.com/blakeblackshear/frigate/pull/1814

Is this in master or somewhere else? Not entirely sure how to deploy a test build with home assistant

**Comment by blakeblackshear** [comment]
As long as this builds ok for me and doesn't impact the RPi builds, I plan to merge these changes.

**Comment by TheStigh** [comment]
@valdeira888 

Are you using Odroid-N2+ (Home Assistant Blue) ?
Got 3 x HD cameras and CPU usage are 35%....

**Comment by dbuezas** [comment]
@valdeira888 the hwaccel_args you posted make ffmpeg crash for me (odroid n2+ and usb coral). Do you have any current tricks to make hw acceleration work ?
Thanks!

**Comment by seanblanchfield** [comment]
Adding some results from a bit of testing, in case it helps someone figure this out. On a ODROID N2+ (Home Assistant Blue), when I try to run Frigate with:
```
ffmpeg:
  hwaccel_args:  -hwaccel drm -hwaccel_device /dev/dri/renderD128
```

I get errors in the Frigate log like the following:
```
[2022-06-30 11:25:24] frigate.video                  ERROR   : driveway: Unable to read frames from ffmpeg process.
[2022-06-30 11:25:24] frigate.video                  ERROR   : driveway: ffmpeg process is not running. exiting 
```
I dug a little deeper by SSHing into Home Assistant and getting a shell in the Frigate container:
```bash
docker exec -it addon_ccab4aaf_frigate bash
```
I used `ps` to find the `ffmpeg` commands that were running and tried running it interactively. 
``` bash
ffmpeg -hide_banner -loglevel warning -hwaccel drm -hwaccel_device /dev/dri/renderD128 -avoid_negative_ts make_zero -fflags nobuffer+genpts+discardcorrupt -flags low_delay -strict experimental -rtsp_transport udp -use_wallclock_as_timestamps 1 -i rtsp:/[RTSP_URL] -f segment -segment_time 10 -segment_format mp4 -reset_timestamps 1 -strftime 1 -c copy -an /media/frigate/hwacceltest.mp4
```
(NOTE: I redacted my RTSP url above, and I modified the output location so I could access the output file easily from my desktop).

This `ffmpeg` command did not crash, and continued running using about 5% CPU (which incidentally might be higher CPU than the regular `ffmpeg` commands without the `hwaccel` arguments). However, the output file did not seem to be a valid video file, as I could not open it in VLC or any other editor.

I think the `-hwaccel drm -hwaccel_device /dev/dri/renderD128` arguments produce invalid video output, which leads to the "**Unable to read frames from ffmpeg process**" error in frigate. I think Frigate responds to this by killing the `ffmpeg` process, which is indicated by the "**ffmpeg process is not running. exiting**" error line. I think we can see this in the code in [video.py:160](https://github.com/blakeblackshear/frigate/blob/93cd973e59ae854382841d1256a97d5b5be733b0/frigate/video.py#L160).
